plugins/woocommerce/includes/admin/helper/class-wc-helper-subscriptions-api.php

```diff
@@ -136,9 +164,83 @@ public static function deactivate( $request ) {
 			wp_send_json_error(
 				array(
 					'message' => __( 'There was an error deactivating your subscription. Please try again.', 'woocommerce' )
-				)
+				),
+				400
+			);
+		}
+	}
+
+	/**
+	 * Install a WooCommerce.com produc
+	 */
+	public static function install( $request ) {
+		$product_key = $request->get_param('product_key');
+		$subscriptions = WC_Helper::get_subscription( $product_key );
+
+		if ( empty( $subscriptions ) ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'We couldn\'t find this subscription.', 'woocommerce' )
+				),
+				404
+			);
+		}
+
+		if ( $subscriptions['expired'] === true ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'This subscription has expired.', 'woocommerce' )
+				),
+				402
 			);
 		}
+
+		if ( $subscriptions['maxed'] === true ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'All licenses for this subscription are already in use.', 'woocommerce' )
+				),
+				402
+			);
+		}
+
+		$activation_success = WC_Helper::activate_helper_subscription( $product_key );
+		if ( ! $activation_success ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'We couldn\'t activate your subscription.', 'woocommerce' )
+				),
+				400
+			);
+		}
+
+		$product_id = $subscriptions['product_id'];
+
+		// Delete any existing state for this product.
+		$state = WC_WCCOM_Site_Installation_State_Storage::get_state( $product_id );
+		if ( $state !== null ) {
+			WC_WCCOM_Site_Installation_State_Storage::delete_state( $state );
+		}
+
+		// Run the installation.
+		$installation_manager = new WC_WCCOM_Site_Installation_Manager( $product_id, $product_id );
```

Comment:
Product installation, though on average taking just a few seconds, can sometimes take more than. 

The installation state is used to ensure that no parallel installation can be initiated while there is already an installation in progress, and the idempotency key ensures that a client can not manage an installation initiated by another client. Neglecting these two protection mechanisms can be really harmful.

Reply:
> The installation state is used to ensure that no parallel installation can be initiated while there is already an installation in progress, and the idempotency key ensures that a client can not manage an installation initiated by another client. Neglecting these two protection mechanisms can be really harmful.

I used this to work around a bug with state not being removed after the product is deleted, but I can take a look at the underlying bug.

Reply:
@bgrgicak The state is supposed to be left until it is intentionally deleted, for tracking installations that have been performed. But it's fine to clear the state to repeat the installation. 

You can get the current state, confirm install is not currently in progress, get the idempotency key and call `\WC_WCCOM_Site_Installation_Manager::reset_installation` to delete the state.

Reply:
https://github.com/woocommerce/woocommerce/pull/40854

Reply:
@bgrgicak Just been looking at this with Cem, and looks like this is no different in https://github.com/woocommerce/woocommerce/pull/40854. Is that right?

Reply:
It's [updated in #40854](https://github.com/woocommerce/woocommerce/pull/40854/files#diff-4cf137656b8d29bec4423f31fb2781a2d67d2866f172c2620ae3a5d4a2d75294R238-R274).
- That PR uses `$state->get_last_step_status()` to confirm an installation isn't in progress before resetting.
- Reset uses ``$installation_manager->reset_installation();` instead of just removing cache.
- `$installation_manager->run_installation` has a try/catch to prevent fatals